src/Streamly/Internal/Data/Fold.hs

```diff
@@ -1445,40 +1407,45 @@ demuxDefaultWith f kv (Fold dstep dinitial dextract) =
                         if Map.size runMap1 == 0
                         then fDone doneMap1
                         else Partial $ fPartial doneMap1 runMap1
-            Done _ -> error "Bug: demuxDefaultWith: Done fold"
+            Done _ -> error "Bug: demuxWith: Done fold"
 
     step (DemuxMapAndDefault dacc doneMap runMap) a = do
-        let (k, a1) = f a
+        let k = f a
+            finished = Map.lookup k doneMap
         case Map.lookup k runMap of
-            Nothing -> do
-                res <- dstep dacc (k, a1)
-                return
-                    $ Partial
-                    $ case res of
-                          Partial s -> DemuxMapAndDefault s doneMap runMap
-                          Done b -> DemuxOnlyMap b doneMap runMap
-            Just fld ->
-                runFold
+            Nothing
+                | isNothing finished ->
+                    runFold
+                    (DemuxMapAndDefault dacc)
+                    (Partial . DemuxOnlyDefault dacc)
+                    doneMap runMap fd k a
+                | otherwise ->
+                    return (Done doneMap)
+            Just fld
+                | isNothing finished ->
+                    runFold
                     (DemuxMapAndDefault dacc)
                     (Partial . DemuxOnlyDefault dacc)
-                    doneMap runMap fld k a1
+                    doneMap runMap fld k a
+                | otherwise ->
+                    return (Done doneMap)
```

Comment:
Since it is already running, it will not be found in the doneMap, so we will always terminate here.

We need a test for demux.